### ISSUE TITLE
Tests are failing

### ISSUE THREAD

**Issue by pkernevez** [vladmihalcea/tsid#15]
I'm facing a concurrency issue, I checkout the repo and run maven:
```
[ERROR] Failures:
[ERROR] io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode
[ERROR]   Run 1: CollisionTest.testCollisionMultipleThreadsPerNode:94 Collisions detected! expected:<0> but was:<1>
[ERROR]   Run 2: CollisionTest.testCollisionMultipleThreadsPerNode:94 Collisions detected! expected:<0> but was:<1>
```

Are tests supposed to work ?
Could it be related to Mac M1 hardware architecture ?

Full trace:
```
> mvn clean install                                                                                                                                                                                 (base)
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------< io.hypersistence:hypersistence-tsid >-----------------
[INFO] Building hypersistence-tsid 2.1.2-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- clean:3.2.0:clean (default-clean) @ hypersistence-tsid ---
[INFO] Deleting /Users/pke/SCM/Git/Shift/hypersistence-tsid/target
[INFO]
[INFO] --- resources:3.3.1:resources (default-resources) @ hypersistence-tsid ---
[INFO] skip non existing resourceDirectory /Users/pke/SCM/Git/Shift/hypersistence-tsid/src/main/resources
[INFO]
[INFO] --- compiler:3.11.0:compile (default-compile) @ hypersistence-tsid ---
[INFO] Changes detected - recompiling the module! :source
[INFO] Compiling 1 source file with javac [debug target 8] to target/classes
[WARNING] bootstrap class path not set in conjunction with -source 8
[WARNING] source value 8 is obsolete and will be removed in a future release
[WARNING] target value 8 is obsolete and will be removed in a future release
[WARNING] To suppress warnings about obsolete options, use -Xlint:-options.
[INFO] /Users/pke/SCM/Git/Shift/hypersistence-tsid/src/main/java/io/hypersistence/tsid/TSID.java: /Users/pke/SCM/Git/Shift/hypersistence-tsid/src/main/java/io/hypersistence/tsid/TSID.java uses or overrides a deprecated API.
[INFO] /Users/pke/SCM/Git/Shift/hypersistence-tsid/src/main/java/io/hypersistence/tsid/TSID.java: Recompile with -Xlint:deprecation for details.
[INFO]
[INFO] --- resources:3.3.1:testResources (default-testResources) @ hypersistence-tsid ---
[INFO] skip non existing resourceDirectory /Users/pke/SCM/Git/Shift/hypersistence-tsid/src/test/resources
[INFO]
[INFO] --- compiler:3.11.0:testCompile (default-testCompile) @ hypersistence-tsid ---
[INFO] Changes detected - recompiling the module! :dependency
[INFO] Compiling 16 source files with javac [debug target 8] to target/test-classes
[WARNING] bootstrap class path not set in conjunction with -source 8
[WARNING] source value 8 is obsolete and will be removed in a future release
[WARNING] target value 8 is obsolete and will be removed in a future release
[WARNING] To suppress warnings about obsolete options, use -Xlint:-options.
[INFO]
[INFO] --- surefire:3.2.2:test (default-test) @ hypersistence-tsid ---
[INFO] Using auto detected provider org.apache.maven.surefire.junit4.JUnit4Provider
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.hypersistence.tsid.TsidTest
[INFO] Tests run: 25, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.078 s -- in io.hypersistence.tsid.TsidTest
[INFO] Running io.hypersistence.tsid.TsidFormatTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s -- in io.hypersistence.tsid.TsidFormatTest
[INFO] Running io.hypersistence.tsid.TsidFactory01024Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.074 s -- in io.hypersistence.tsid.TsidFactory01024Test
[INFO] Running io.hypersistence.tsid.TsidFactory00256Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.091 s -- in io.hypersistence.tsid.TsidFactory00256Test
[INFO] Running io.hypersistence.tsid.TsidFactoryTest
[WARNING] Tests run: 25, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 0.048 s -- in io.hypersistence.tsid.TsidFactoryTest
[INFO] Running io.hypersistence.tsid.TestSuite
[ERROR] Tests run: 94, Failures: 1, Errors: 0, Skipped: 2, Time elapsed: 1.610 s <<< FAILURE! -- in io.hypersistence.tsid.TestSuite
[ERROR] io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode -- Time elapsed: 0.484 s <<< FAILURE!
java.lang.AssertionError: Collisions detected! expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode(CollisionTest.java:94)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

[INFO] Running io.hypersistence.tsid.TsidBaseNTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s -- in io.hypersistence.tsid.TsidBaseNTest
[INFO] Running io.hypersistence.tsid.TsidFactory16384Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.021 s -- in io.hypersistence.tsid.TsidFactory16384Test
[INFO] Running io.hypersistence.tsid.TsidFactory00001Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.524 s -- in io.hypersistence.tsid.TsidFactory00001Test
[INFO] Running io.hypersistence.tsid.CollisionTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.444 s <<< FAILURE! -- in io.hypersistence.tsid.CollisionTest
[ERROR] io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode -- Time elapsed: 0.290 s <<< FAILURE!
java.lang.AssertionError: Collisions detected! expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode(CollisionTest.java:94)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

[INFO] Running io.hypersistence.tsid.IncrementalTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.031 s -- in io.hypersistence.tsid.IncrementalTest
[INFO] Running io.hypersistence.tsid.TsidFactory04096Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.019 s -- in io.hypersistence.tsid.TsidFactory04096Test
[INFO] Running io.hypersistence.tsid.TsidFactory00064Test
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.022 s -- in io.hypersistence.tsid.TsidFactory00064Test
[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR] io.hypersistence.tsid.CollisionTest.testCollisionMultipleThreadsPerNode
[ERROR]   Run 1: CollisionTest.testCollisionMultipleThreadsPerNode:94 Collisions detected! expected:<0> but was:<1>
[ERROR]   Run 2: CollisionTest.testCollisionMultipleThreadsPerNode:94 Collisions detected! expected:<0> but was:<1>
[INFO]
[INFO]
[ERROR] Tests run: 185, Failures: 1, Errors: 0, Skipped: 2
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.556 s
[INFO] Finished at: 2024-05-09T19:13:26+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.2:test (default-test) on project hypersistence-tsid: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/pke/SCM/Git/Shift/hypersistence-tsid/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Comment by pkernevez** [comment]
The failing test is using TSID.Factory.THREAD_LOCAL_RANDOM_FUNCTION as our project.

**Comment by pkernevez** [comment]
And I can't understand what this test is doing.
It says: 	testCollisionMultipleThreadsPerNode, but all the threads are using a different instance of the factory with exactly the same parameters (no withNode( xxx) ).

For what I understand, they are supposed to share the same Factory instance. If not the code is not working because ThreadLocalRandom.current().nextInt() wont provide you guaranteed different ranges.

**Comment by vladmihalcea** [comment]
The test does not fail for me, but there can be collisions. The test only checks the probability of collision using that ThreadLocalRandom.

It could be that your M1 architecture makes the test fail. Anyway, it should be ignored as was used more for stress testing.

**Comment by pkernevez** [comment]
They are failing often on my laptop: about 35% of the time
24 times with 1 collision
9 times it was with 2 collisions
1 time with 3 collisions
1 time with 4 collisions

```
$  for i in $(seq 1 100); do mvn test -Dtest=CollisionTest#testCollisionMultipleThreadsPerNode | grep CollisionTest.testCollisionMultipleThreadsPerNode:;done | wc -l
   35
```

**Comment by vladmihalcea** [comment]
Fixed along with #5
